airbyte-integrations/connectors/source-braintree/source_braintree/streams.py

```diff
@@ -0,0 +1,215 @@
+from datetime import datetime
+from typing import Any, Iterable, List, Mapping, Optional, Union
+
+import backoff
+import braintree
+import pendulum
+from airbyte_cdk.models import SyncMode
+from airbyte_cdk.sources.streams.core import Stream
+from braintree.attribute_getter import AttributeGetter
+from source_braintree.schemas import Customer, Discount, Dispute, MerchantAccount, Plan, Subscription, Transaction
+from source_braintree.spec import BraintreeConfig
+
+
+class BraintreeStream(Stream):
+    def __init__(self, config: BraintreeConfig):
+        self._start_date = config.start_date
+        self._gateway = BraintreeStream.create_gateway(config)
+
+    @staticmethod
+    def create_gateway(config: BraintreeConfig):
+        return braintree.BraintreeGateway(braintree.Configuration(**config.dict()))
+
+    @abstractproperty
+    def model(self):
+        """
+        Pydantic model to represent catalog schema
+        """
+
+    @abstractmethod
+    def get_items(self, start_date: datetime):
+        """
+        braintree SDK gateway object for items list
+        """
+
+    def get_json_schema(self):
+        return self.model.schema()
+
+    def stream_slices(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_state: Mapping[str, Any] = None,
+    ) -> Iterable[Optional[Mapping[str, Any]]]:
+
+        current_datetime = pendulum.utcnow()
+        if sync_mode == SyncMode.full_refresh:
+            return [{self.cursor_field or "start_date": self._start_date or current_datetime}]
+        stream_state_start_date = stream_state.get(self.cursor_field)
+        if stream_state_start_date:
+            stream_state_start_date = pendulum.parse(stream_state_start_date)
+        start_date = stream_state_start_date or self._start_date or current_datetime
+        return [{self.cursor_field: start_date}]
+
+    def get_updated_state(
+        self,
+        current_stream_state: Mapping[str, Any],
+        latest_record: Mapping[str, Any],
+    ):
+        next_state = latest_record.get(self.cursor_field)
+        current_state = current_stream_state.get(self.cursor_field)
+        current_state = pendulum.parse(current_state) if current_state else next_state
+        return {self.cursor_field: max(current_state, next_state).strftime("%Y-%m-%d %H:%M:%S")}
+
+    @staticmethod
+    def get_json_from_resource(resource_obj: Union[AttributeGetter, List[AttributeGetter]]):
+
+        if isinstance(resource_obj, list):
+            return [obj if not isinstance(obj, AttributeGetter) else BraintreeStream.get_json_from_resource(obj) for obj in resource_obj]
+        obj_dict = resource_obj.__dict__
+        return {
+            attr: obj_dict[attr]
+            if not isinstance(obj_dict[attr], (AttributeGetter, list))
+            else BraintreeStream.get_json_from_resource(obj_dict[attr])
+            for attr in obj_dict
+            if not attr.startswith("_")
+        }
```

Comment:
```suggestion
        return {
            attr: obj_dict[attr] if not isinstance(obj_dict[attr], (AttributeGetter, list)) else BraintreeStream.get_json_from_resource(obj_dict[attr])
            for attr in obj_dict if not attr.startswith("_")
        }
```

When this dict creation is on 5 lines it's really hard to read it.

Reply:
Also, why are you using `if not ... else`? Why don't you use `if ... else`? Like this:
```
{
    attr: BraintreeStream.get_json_from_resource(obj_dict[attr]) if isinstance(obj_dict[attr], (AttributeGetter, list)) else obj_dict[attr]
}
```

Reply:
I think it's better not to use dict comprehension at all here and write normal `for` loop. It's really hard to read this dict comprehension.

Reply:
fixed